File: C programs/437. Path Sum III.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

long long dfs(struct TreeNode* node, long long target) {
    if (!node) {
        return 0;
    }

    long long count = 0;
    if (node->val == target) {
        count++;
    }

    count += dfs(node->left, target - node->val);
    count += dfs(node->right, target - node->val);

    return count;
}

long long pathSum(struct TreeNode* root, long long targetSum) {
    if (!root) {
        return 0;
    }

    long long total_count = dfs(root, targetSum);
    total_count += pathSum(root->left, targetSum);
    total_count += pathSum(root->right, targetSum);

    return total_count;
}

File: other languages/437. Path Sum III.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        if not root:
            return 0
        total = self.dfs(root, targetSum)
        total += self.pathSum(root.left, targetSum)
        total += self.pathSum(root.right, targetSum)
        return total

    def dfs(self, root, sum1):
        count = 0
        if not root:
            return count
        if sum1 == root.val:
            count += 1
        count += self.dfs(root.left, sum1 - root.val)
        count += self.dfs(root.right, sum1 - root.val)
        return count
